app/scripts/hooks/usePromoReport.ts

```diff
@@ -0,0 +1,101 @@
+import { Conference } from 'conference';
+import { find } from 'lodash';
+import { PromotionReport } from 'promotionReport';
+import { useEffect, useMemo, useState } from 'react';
+import { PromoReportService } from '../services/promoReportService';
+import { PromoRegistration } from './usePromoRegistrationList';
+
+let activeRequest: Promise<PromotionReport> | null = null;
+
+export const usePromoReport = ({
+  conference,
+  reportId,
+  promoReportService,
+}: {
+  conference: Conference;
+  reportId: string | null;
+  promoReportService: PromoReportService;
+}): {
+  // The loaded promo report, which will be null while it is being loaded
+  report: PromotionReport | null;
+
+  // The promo registrations associated with the report
+  promoRegistrations: Array<PromoRegistration>;
+
+  // Indicates whether the promo report is currently loading from the server
+  loading: boolean;
+
+  // Reload the promo report
+  refresh: () => void;
+} => {
+  const [report, setReport] = useState<PromotionReport | null>(null);
+  const [loading, setLoading] = useState(false);
+
+  const refresh = () => {
+    setReport(null);
+    setLoading(true);
+
+    if (reportId === null) {
+      // If the report is null, don't load and keep the report null
+      setLoading(false);
+      activeRequest = null;
+    } else {
+      // To avoid getting confused if there are multiple requests in progress, track the most recent
+      // request and ignore the response if a request comes back that isn't the most recent request.
+      const currentRequest = promoReportService.loadPromoReport(
+        conference.id,
+        reportId,
+      );
+      activeRequest = currentRequest;
+      currentRequest.then((report) => {
+        if (activeRequest === currentRequest) {
```

Comment:
Won't `activeRequest` always equal `currentRequest` as the variable `activeRequest` is set to `currentRequest` 2 lines up?

Reply:
No, because it's defined outside of the hook. The idea is that `activeRequest` will keep being overwritten by each request and will hold the most recent request, while `currentRequest` will remain constant. After a request asynchronously finishes, it can check whether a different request started after it, in which case its response will be discarded in favor of the more recent request. It's a complex flow, so sorry if my explanation doesn't make sense. I'm also in the process of switching from a global `activeRequest` variable to a React ref so that multiple active `usePromoReport` hooks won't overwrite each others' `activeRequest`.